### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by matthewp** [amodrojs/amodro-lifecycle#1]
Hi @jrburke, I maintain a few advanced loader extensions based on the previous es6 spec which this seems to take some inspiration from.  I'm interested in trying to implement some of these using your skeleton.  I like that it is based on a constructor so inheritance will work from the start (something the es6 spec still doesn't have). 

Is the best way to test it out to fork and add an implementation to `impl/` and a corresponding test in `test/`?

Or perhaps this isn't quite ready for feedback yet?

**Comment by jrburke** [comment]
I am hoping to move the stuff out of impl into its own repo soon, so I suggest the same in this case, using a separate repo that just pulls in a version of the the lifecycle.js file.

I just had the impl thing in here at first just to shake out basic operation, and did not want to keep copying over lifecycle as I did changes. For instance, this week I did a few new passes on the cycle resolution as I got the AMD impl passing the requirejs tests.

I have almost all of the requirejs tests passing now, except for the undef/paths fallbacks/error pathway cases because I have not gotten to those extensions yet. However, feeling pretty good now about the core operation of lifecycle. I want to do some other more minor changes then split the impl/amd out into just the amodro repo, and version rev lifecycle, then start reving lifecycle separately as more usage points to changes.

I hope to complete that this weekend, so you might want to wait for that, or just monitor commits to lifecycle here as I finish out that work. But using a separate repo for any loader work on top of lifecycle is the way to go, as I will be doing the same.

Closing as a discussion ticket, but feel free to continue discussion here, just going to keep open tickets for things that might need code changes/docs/etc.

**Comment by matthewp** [comment]
Pick up from [our other discussion](https://github.com/whatwg/loader/issues/50#issuecomment-111594085) I think this is a bug.  Consider:

``` js
loader.useUnnormalized("rafPolyfill?has(requestAnimationFrame)").then(raf){
  // use raf
});
```

This will try to synchronously normalize the string "rafPolyfill?has(requestAnimationFrame)" which needs to be normalized async.

Sense depend() is async, you should be able to use it's result as the new id.

Change: https://github.com/amodrojs/amodro-lifecycle/blob/master/lifecycle.js#L213

to:

``` js
      .then(function(normalizedIds) {
        fslog(factoryTree, 'useUnnormalized.then: ' + id + ', ' + refId);
        try {
          normalizedId = this.top.normalize(normalizedIds[0], refId);
```

would fix it.

This is a bit of a silly example but I have a real extension that needs this behavior if you want me to go into it.

**Comment by matthewp** [comment]
Unrelated (and I'll ask somewhere else if it's more appropriate), but do you / how do you plan to support allowing users to add extensions to your loader?  For example, you have and AMD loader and someone wants to add an extension that enables loading ES modules (by compiling in Babel or whatever).

Of course you could subclass the AMD loader to add the functionality, this is a great code solution but not user-friendly.  Someone has to maintain a project which combines all of the needed extensions.  I've been down this road with SystemJS and it's not great for community-building.

What would be really nice is if the user could specify a list of extensions to include.  I'm not sure exactly how this should work, but it would enable 1 community (amodro or whatever) with a nice ecosystem built around it.

Maybe plugins are your solution here, my apologies as I'm not very familiar with requirejs plugins.

**Comment by matthewp** [comment]
@jrburke So I did a naive implementation of CommonJS (for the browser) [here](https://github.com/matthewp/a-commonjs) to play around with lifecycle.js. Here's a few points of feedback you may or may not find useful:
1. **evaluate** seems misnamed.  For AMD and Node CommonJS it makes sense but doesn't for ES6 or Browser CommonJS because you are not going to eval in either of those cases.  I would instead call it **parse** because, thinking in terms of ES6 modules, you parse for dependencies before evaluation.  fetch -> translate -> parse -> instantiate.  For ES6 and Browser CommonJS the factory function is where eval will actually happen.
2. It would be great if there were a metadata object (like in the es6 spec) that got passed to every hook after normalize.  For the case of CommonJS I needed a map of unnormalized to normalized dependency ids. I'd also want this info in a build scenario so putting this type of extra info on a metadata object would be a good convention.

There was a third thing that has escaped me now, I'll post if I remember.

Next I'll probably try implementing an npm extension on top of the commonjs loader, this is the one I was talking about that needed an async normalize, so we'll see if **depend** serves my needs (I think it probably will).

**Comment by jrburke** [comment]
Sorry for the late reply, was trying to get more code together. For example, separated the AMD variant out and put it under amodro-base. I did an example "poly!" plugin here:

https://github.com/amodrojs/amodro-base/tree/master/test/plugin-poly

So in that model, the plugin does not need to normalize the ID, the ID is already in a normalized form. Although that example could implement a normalize if it wanted to, just to save the base loader from doing unnecessary normalization work. And the loader plugin would be loaded before the first normalize() call is called.

The work is really targeted in the fetch, which can set the module value and not do a real fetch, and that will satisfy the definition for that module ID.

If you do not think that sort of approach works for the other real world case you have, knowing more about it would be helpful.

> Unrelated (and I'll ask somewhere else if it's more appropriate), but do you / how do you plan to support allowing users to add extensions to your loader?

There are two kinds of extensions:
- loader plugins, which can be loaded on demand, as needed. The poly! one is an example of that type. A text! plugin or coffeescript! plugin, or other transpiler type loader plugins are other examples.
- generic overrides to the lifecycle steps. These need to be in place before loading starts, given the generic nature of them. So I expect using a build tool to construct them. It could be as simple as a concatenation. For amodro-base, where I have the parts for the AMD variant on top of amodro-lifecycle, it has a build process, in the build/build.js folder. It actually allows insertion of files inside the "main.js" base file. 

That build proces is a bit basic and probably not very flexible at the moment, but illustrates the general idea: if fancier generic overrides need to be bundled up with the loader so that they are in place before loading of modules begin. It means properly sequencing the overrides that are desired, but it gives a fair bit of flexibility.

> 1. evaluate seems misnamed. 

I appreciate the point, but given that it can be more than just parse in some cases, I would prefer a more general name than parse. Loader plugins could do extra work in there besides parsing. Maybe analyze? But given there can be side effects, not just a passive read, not sure if that is the right word either.

> 1. It would be great if there were a metadata object (like in the es6 spec) that got passed to every hook after normalize.

Since Lifecycle is set up to be "subclassed", with prototype properties, I prefer to encourage using the prototype for storing data between the steps, if coordination is needed across disparate plugins or lifecycle overrides.

In other experiments, I ran into trouble when state information like that had to make it through multiple overrides, and it creates more of a burden on overrides to properly pass it around. It felt cleaner to treat the lifecycle steps as more functional steps.

For state that just needs to be private for a given loader plugin over multiple lifecycle steps (and not shared with other loader plugins or overrides), keeping that state as private objects inside the loader plugin module, with the data indexed by normalizedId felt more appropriate.

**Comment by matthewp** [comment]
Thanks for your reply.

> If you do not think that sort of approach works for the other real world case you have, knowing more about it would be helpful.

The use case is an extension that loads npm modules; we need to load a `package.json` in order to properly normalize the name (`foo` becomes `foo@1.0.0#lib/foo/index`) otherwise you won't have distinct module names. `depend` is a good solution for this, but you're not using the result of depend [here](https://github.com/amodrojs/amodro-lifecycle/blob/master/lifecycle.js#L225), otherwise it would be good.

> Maybe analyze? 

I do think that's better. My concern with the `evaluate` hook is that it's very AMD-centric and if the goal is to gain adoption it might be better to name things ES-centric.  If that's not part of the goal then it doesn't matter, it doesn't bother me much.

>  I prefer to encourage using the prototype for storing data between the steps, if coordination is needed across disparate plugins or lifecycle overrides.

Sorry, I didn't properly explain my need for metadata. The metadata is per module, not per loader. For example you might have a loader that supports AMD, CommonJS and ES modules, you could use a metadata object to assign `{ "format": "amd"}` to the module so that it can be used by a build-tool for transpiling (say you are transpiling everything to amd for bundling).

It does place some burden on overrides to pass it around, but the same burden exists for the arguments you have now. I think that's why the "load" object in the ES6 spec was a good idea, every hook after normalize was given a "load" object that looked like: `{name: "foo", address: "http://example.com/foo.js", metadata: {} }`.  So overrides only have to remember to pass the `load` object.

I do understand wanting to make the hooks pure, so perhaps having 1 impure hook where this type of information can be assigned (post-evaluate I think) is the way to go.

I do have more thoughts on the different types of extensions, but I don't think this is the appropriate place for that as it's a loader question not a lifecycle question, so I'll save those for when your AMD loader is ready for feedback.

**Comment by jrburke** [comment]
> The use case is an extension that loads npm modules; we need to load a package.json in order to properly normalize the name (foo becomes foo@1.0.0#lib/foo/index) otherwise you won't have distinct module names. depend is a good solution for this, but you're not using the result of depend here, otherwise it would be good.

For that use case, how does the loader know that it should look for a package.json for 'foo'? Does it just assume all top level ID segments, so 'foo' in the example, or 'bar' in 'bar/something', are package directories? Or is some config involved? How does it know to look for foo@1.0.0 vs foo@2.0.0? I think there is way for it to work, but want to be sure I understand before trying an example.

> My concern with the evaluate hook is that it's very AMD-centric and if the goal is to gain adoption it might be better to name things ES-centric.

Thinking more about it, parse seems better than analyze, and probably fits better with the other hooks; between translate and depend a parse makes sense. It can just be that a script eval in the legacy script case could also happen. The main goal is for something to end up in the registry, and not something necessarily evaluated, so perhaps parse is the way to go.

> So overrides only have to remember to pass the load object.

I can see the attraction of using a property bag as the only argument, maybe to hope to allow future expansions without needing to modify arguments. It feels a bit too free-form to me though. It gets hard to know what the contract is for a given step. It is certainly a known school of thought on design, if that was the thing to close out getting an ES loader spec done, then fine. 

I would like to give the explicit args some more time though, get more feedback on it, particularly since I do not think it is as flexible as storing the data on the loader object itself, example following.

I was also probably not as clear in my earlier response, I meant a per loader object that stored data by module ID. So in the case where you wanted to store the module type, create a `this.moduleFormat = {}` in the loader's constructor function, and in the translate phase, when it discovers the module type, set it via `this.moduleFormat[normalizedId] = 'cjs'`, and that can then be consulted in perhaps the evaluate phase.

There is something similar I do for the amodro-es variant of amodro-base, since it needs to adapt the exports based on if an es module asks for an AMD module. It tracks that info in `this.esType`. The nice thing about storing it on the loader (but keyed off module ID) means it can be used in the lifecycle overrides for non-step functions, like getModule, which are APIs that may not get a `load` object passed to them.

That said, I just realized that the loader plugin support in amodro-base creates a proxy object for the loader, so as not to expose all of the loader internals to the loader plugin, and the proxy object would not necessarily know what to allow in the proxy for this freeform data. Perhaps it is enough then for the loader to provide a `getData(normalizedId)` function on its prototype that would handle storing the grab bags per ID. I'll take a look at adding that.

**Comment by matthewp** [comment]
> For that use case, how does the loader know that it should look for a package.json for 'foo'? Does it just assume all top level ID segments, so 'foo' in the example, or 'bar' in 'bar/something', are package directories? Or is some config involved? How does it know to look for foo@1.0.0 vs foo@2.0.0? I think there is way for it to work, but want to be sure I understand before trying an example.

If there is a parent and the parent has an npm-like name it looks in the parent's package.json for its dependencies. Otherwise it will look in the top-level package.json for dependencies.  So in the case where you do: `loader.useUnnormalized("foo")` it would look in the top level package.json for a dependency called "foo".

So the way depend is set up this will work if [this line](https://github.com/amodrojs/amodro-lifecycle/blob/master/lifecycle.js#L225) were to use the result of depend.  I can't think of a reason not to do that anyways, isn't that the reason depend is called?

> Perhaps it is enough then for the loader to provide a getData(normalizedId) function on its prototype that would handle storing the grab bags per ID. I'll take a look at adding that.

I would make the result mutable, otherwise it's going to have to internally do a merge of some sort.

``` js
var metadata = loader.getData("foo");
metadata.cacheKey = Math.random(); // for a cache-busting extension
```

**Comment by jrburke** [comment]
I pushed a 0.2.1 of lifecycle that includes renaming evaluate to parse, and a getData() with mutable object returned, thanks for the suggestions.

On the package.json example, I did a pass at that in this test example:

https://github.com/amodrojs/amodro-base/blob/master/test/package-json-override/test.html

I do not think using the result of depend() in useUnnormalized will work out. The goal of depend is to allow inspection of unnormalized IDs, but the end result of that inspection is still unnormalized IDs. If some depend overrides decide to normalize, it could lead to a double-normalization, which if config like map config is in play will result in incorrect results.

But I think the issue is solvable, hopefully that test shows one way. If you want to see different ways or if there is more nuance to the use case, feel free to add more detail.

**Comment by matthewp** [comment]
Thanks for taking so much time listening to my weird needs.  I do think getData() will be important for tooling so I'm glad you added that in.

I guess I misunderstood the purpose of depend. It does make sense that normalizing at this step could cause bad things if using something like map.  So you're right not to use it there. 

I guess for my use case I can use depend to load the package.jsons that I need so that they're available for inspection in normalize. I'll give that a go when I have some time.

What was the use-case for depend that you had in mind when you created it that would need it to be asynchronous?

**Comment by jrburke** [comment]
To allow lifecycle overrides to make sure they have all they need to properly normalize the dependencies, they could also explicitly add or remove dependencies from the parsed result. 

But mainly it grew out of making sure loader plugins could be supported: if a 'text!something.html' dependency is seen, the lifecycle override can make sure the 'text' module is fetched and available before normalization of 'text!something.html' is attempted. This sort of package.json fetching also qualifies as a depend hook use case.

This design allows the other lifecycle steps to be synchronous, in particular normalize and locate, which are also usually desired to be synchronous on a module meta object.

(side note: I will be unresponsive/super laggy response for a couple of weeks as I will not have good keyboard access)

**Comment by matthewp** [comment]
Hey @jrburke, my day job has been keeping me busy so I haven't had much time to reflect on Amodro until now.

I went ahead and implemented an NPM loader (with CommonJS support) that is based on blocking in `depend` [which you can see here](https://github.com/matthewp/a-commonjs).  It's not ideal because it crawls the entire package.json dependency tree rather than waiting for modules to be required but I'm sure some small adjustments would make this possible as well.

Suffice to say I'm hopeful/optimistic that this will work with advanced Loader extensions which is great.  I'd like to comment on how to make the Amodro ecosystem vibrant, my experience with SystemJS was that making a new Loader project for each set of extensions wanted led to there being no ecosystem to speak of so I'd like to discuss how Amodro can do better, but I'll leave that for another time (and another repo).

Thanks for your hard work in setting up this excellent spec.  One question, how confident are you that this same lifecycle could be bolted on top of the whatwg loader if/when that finally comes to browsers?  I'm thinking maybe you just need to override `import` and do everything that happens in Lifecycle's `load` there.

**Comment by jrburke** [comment]
> One question, how confident are you that this same lifecycle could be bolted on top of the whatwg loader if/when that finally comes to browsers? I'm thinking maybe you just need to override import and do everything that happens in Lifecycle's load there.

It really depends on what ends up in the whatwg loader spec. One of the main goals of this experiment was to show how this sort of lifecycle would work, so that it is considered in some similar form for the whatwg loader spec. Hard to know though given the communication style for that spec.
